### PR TITLE
[TS] LPS-157425 (Initial Review)

### PR DESCRIPTION
**Pull request by ericyanLr**
Hi Team,

**Context**
This PR was created with guidance from [PTR-3915](https://issues.liferay.com/browse/PTR-3915) and will serve as the initial PR for the Search Team and Headless Team to review.

**Goal**
The goal of this PR is to restore the filtering functionality of Headless API, which was partially broken after the Breaking Change that was introduced by [LPS-152937](https://issues.liferay.com/browse/LPS-152937)

**PR**
This PR basically introduces 2 types of changes:
- Headless: To no longer utilize *_sortable fields for filtering String fields
- Search: Update mapping to introduce multi-field keyword fields to replace the previous *_sortable fields

Note: For the multi-field keyword fields, I ended up setting the normalizer to lowercase. 
- See commit (https://github.com/liferay-search/liferay-portal/commit/205dabdd6496036ef394c0cb158e95b3e84d1de6) for details

Please let me know if you have any questions.
Thanks!

/cc @lipusz @peerkar @4lejandrito

### CODE REVIEW

**Comment by lipusz** [comment]
@ericyanLr Do you intentionally use an old ticket number?

**Comment by lipusz** [comment]
Asking because if it's not having a new LPS and it's not in our Sprint, it won't show up on our boards.

**Comment by joshuacords** [comment]
ci:test:sf

**Comment by ericyanLr** [comment]
@lipusz, can you help clarify what you mean by:
> @ericyanLr Do you intentionally use an old ticket number ([LPS-152937](https://issues.liferay.com/browse/LPS-152937))?
> Asking because if it's not having a new LPS and it's not in our Sprint, it won't show up on our boards.

The main ticket is under: [LPS-157425](https://issues.liferay.com/browse/LPS-157425). 

As for [LPS-152937](https://issues.liferay.com/browse/LPS-152937), this was mentioned in the ticket description to indicate the breaking change that introduced this bug. If there's something I'm missing and require changing on my end, please let me know.

Thanks!

File: modules/apps/portal-odata/portal-odata-api/src/main/java/com/liferay/portal/odata/entity/StringEntityField.java [hunk 1]
```diff
@@ -44,4 +44,25 @@ public StringEntityField(
...
+	 * @param  name the entity field's name
+	 * @param  sortableFieldNameFunction the sortable field name {@code
+	 *         Function}
+	 * @param  filterableFieldNameFunction the filterable field name {@code
+	 *         Function}
+	 * @review
+	 */
+	public StringEntityField(
+		String name, Function<Locale, String> sortableFieldNameFunction,
+		Function<Locale, String> filterableFieldNameFunction) {
```
Comment (joshuacords) [inline modules/apps/portal-odata/portal-odata-api/src/main/java/com/liferay/portal/odata/entity/StringEntityField.java:61]: Might want to alphabetize the function names.

Comment (joshuacords) [reply modules/apps/portal-odata/portal-odata-api/src/main/java/com/liferay/portal/odata/entity/StringEntityField.java:61]: (I hope the IDE Refactor takes care of all instances)

**Comment by joshuacords** [comment]
On my first read through things are looking right. Is there any tests that insure the filtering/sorting of Headless requests? 
@ericyanLr

**Comment by joshuacords** [comment]
Hello @ericyanLr, have you also tested Solr to see if changes are also necessary there?

File: modules/apps/portal-odata/portal-odata-api/src/main/java/com/liferay/portal/odata/entity/StringEntityField.java (hunk 1 above)
Comment (ericyanLr) [reply modules/apps/portal-odata/portal-odata-api/src/main/java/com/liferay/portal/odata/entity/StringEntityField.java:61]: The reason I did that order was to match the parent class: https://github.com/liferay-search/liferay-portal/blob/bc89c7fb837632c72cc7b7754d93eff2ca846085/modules/apps/portal-odata/portal-odata-api/src/main/java/com/liferay/portal/odata/entity/EntityField.java#L46-L49

Also, SF didn't complain haha. Do you know if the preference would be to alphabetize the function names regardless?

**Comment by ericyanLr** [comment]
@joshuacords, regarding:
> On my first read through things are looking right. Is there any tests that insure the filtering/sorting of Headless requests? @ericyanLr

There are tests, but not sure if there's coverage on this particular test case we are trying to handle. Would have to confirm.

**Comment by ericyanLr** [comment]
@joshuacords , regarding:
> have you also tested Solr to see if changes are also necessary there?

I have not tested on Solr and not sure how to determine whether additional changes are necessary. Is that just tested via manual testing?

**Comment by joshuacords** [comment]
@ericyanLr,
You could just test manually. Basically:

1. Build/deploy the Solr 8 connector modules.
2. Remove the ES7 api, impl, spi modules
3. Start bundle
4. Full Reindex
5. Test the same steps and see if the Headless endpoints work using Solr

**Comment by joshuacords** [comment]
> @joshuacords, regarding:
> 
> > On my first read through things are looking right. Is there any tests that insure the filtering/sorting of Headless requests? @ericyanLr
> 
> There are tests, but not sure if there's coverage on this particular test case we are trying to handle. Would have to confirm.

Considering this is a regression already, I think there should be at least some tests in place to make sure this Headless functionality is maintained.

**Comment by ericyanLr** [comment]
Hi @joshuacords ,

**Regarding:**

> have you also tested Solr to see if changes are also necessary there?
> ...
> Considering this is a regression already, I think there should be at least some tests in place to make sure this Headless functionality is maintained.

Before proceeding to polish this PR, I do want to confirm with both teams that this approach is moving towards the right direction. Unless, you feel we need to complete these additional tasks, before the Search team can make a proper evaluation.

This PR was created as an attempt to apply what was discussed/suggested between both teams in [PTR-3915](https://issues.liferay.com/browse/PTR-3915). To me, this PR would serve to allow both teams to evaluate the discussed proposed changes and for both teams to eventually come into an agreement on a final solution.

**Comment by joshuacords** [comment]
Hi @ericyanLr,

The Search Team considers this the best way to deliver the desired behavior.

**Comment by ericyanLr** [comment]
Thanks Josh!

@4lejandrito, can you help take a look at this PR + discussion and let us know what the Headless team thinks about the current proposed approach?

**Comment by 4lejandrito** [comment]
I am ok with the approach.

> > @joshuacords, regarding:
> > > On my first read through things are looking right. Is there any tests that insure the filtering/sorting of Headless requests? @ericyanLr
> > 
> > 
> > There are tests, but not sure if there's coverage on this particular test case we are trying to handle. Would have to confirm.
> 
> Considering this is a regression already, I think there should be at least some tests in place to make sure this Headless functionality is maintained.

@magjed4289 can you shed some light on this? Do we have tests for these cases? How was https://issues.liferay.com/browse/LPS-157425 discovered and marked as regression?

**Comment by ericyanLr** [comment]
Hi @magjed4289,

When you get a chance, can you help take a look at @4lejandrito 's comment above?

**Comment by magjed4289** [comment]
@4lejandrito @ericyanLr @joshuacords 
We test `startsWith` and `contains` when filtering in Object. AFAIK no tests for other modules, but QA of the correspondent teams would maybe have more info.

**Comment by ericyanLr** [comment]
Hi @magjed4289 ,

Would you able to share a reference to one of those tests you mentioned? Also, do you know if those tests are utilizing Elasticsearch to retrieve the filtered resulted?

Doing a quick search, the closest I could find were:
- [FilterParserImplTest.java](https://github.com/liferay/liferay-portal/blob/7.4.3.79-ga79/modules/apps/portal-odata/portal-odata-impl/src/test/java/com/liferay/portal/odata/internal/filter/FilterParserImplTest.java)
- [BaseStructuredContentResourceTestCase.java#L1531-L1564](https://github.com/liferay/liferay-portal/blob/7.4.3.79-ga79/modules/apps/headless/headless-delivery/headless-delivery-test/src/testIntegration/java/com/liferay/headless/delivery/resource/v1_0/test/BaseStructuredContentResourceTestCase.java#L1531-L1564) 
  - testGetSiteStructuredContentsPageWithFilterStringEquals()

**Comment by ericyanLr** [comment]
Resent with follow-up changes: https://github.com/liferay-search/liferay-portal/pull/953

**Pull request closed without merging**